text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3436 @@
+
+Since scoped implementations allow crates to reusably implement external traits on external types, they can be used to provide API extensions that make use of syntactic sugar. For example:
+
+Filename: fruit-comparer/src/lib.rs
+
+```rust
+use apples::Apple;
+use oranges::Orange;
+
+pub use impl PartialEq<Orange> for Apple {
+    fn eq(&self, other: &Orange) -> bool {
+        todo!("Figure out how to compare apples and oranges.")
+    }
+}
+
+pub use impl PartialEq<Apple> for Orange {
+    fn eq(&self, other: &Orange) -> bool {
+        todo!("Figure out how to compare oranges and apples.")
+    }
+}
+```
+
+Filename: src/main.rs
+
+```rust
+use apples::Apple;
+use oranges::Orange;
+
+use fruit_comparer::{
+    impl PartialEq<Orange> for Apple,
+    impl PartialEq<Apple> for Orange,
+};
+
+fn main() {
+    let apple = Apple::new();
+    let orange = Orange::new();
+
+    // Compiles:
+    dbg!(apple == orange);
+    dbg!(orange == apple);
+}
+```
+
+If the type whose API was extended this way later gains the same trait inherently, that is not a problem as the consuming code continues to use `fruit_comparer`'s scoped implementation. However, a warning ([global-trait-implementation-available]) is shown by default to alert the maintainers of each crate of the covering global implementation.
+
+Be careful about literal coercion when using generic traits this way! For example, if a scoped implementation of `Index<isize>` is used and a global `Index<usize>` implementation is added later on the same type, the compiler will *not* automatically decide which to use for integer literal indices between these two.
+
+## Rustdoc documentation changes
+[rustdoc-documentation-changes]: #rustdoc-documentation-changes
+
+### `use` and `impl` keywords
+
+The documentation pages [for the `use` keyword] and [for the `impl` keyword] are adjusted to (very) briefly demonstrate the respective scoped use of `impl Trait for Type`.
+
+[for the `use` keyword]: https://doc.rust-lang.org/stable/std/keyword.use.html
+[for the `impl` keyword]: https://doc.rust-lang.org/stable/std/keyword.impl.html
+
+### `TypeId`
+
+The page for [`TypeId`] gains two sections with the following information:
+
+```markdown
+# `TypeId` and scoped implementations
+
+To make sure that that are no mix-ups between, for example, `HashSet<T>` and
+`HashSet<T as Hash in module>`, any such difference implies distinct `TypeId`s between
+such discretised generics (and that the types are not mutually assignable).
+
+This also affects trait-bounded generic type parameters: If `T` is bounded on `Hash`, then
+`TypeId::of::<T>()` results in distinct `TypeId`s in that context depending on the
+captured implementation.
+
+However, note that `TypeId::of::<T>()` and `TypeId::of::<T as Hash in module>()` are
+always equivalent for one definition of `T`, as `TypeId::of`'s implementation does **not**
+have a `T: Hash` bound!
```

Comment:
`Type as Trait in module` is technically not a type at all, as implementations never "stick" to the outside of a type.
It's a type argument consisting of a type and an inline implementation environment.

This combination, along with any scoped implementations inherited from the surrounding scope, is "baked" into the effective (opaque) type of the generic type parameter as part of monomorphisation.
These opaque types are special-cased in terms of how the baked implementation environment influences the observed `TypeId`:

- If the type parameter is wrapped in an implementation-aware generic (even indirectly), e.g. as `TypeId::of::<HashSet<T>>()`, then the `TypeId` identifies the combination of the discrete type given as part of the type argument and the *full* set of baked implementations.
  (Normally this would happen because the e.g. `HashSet<T>` is used as type argument and *then* has its `TypeId` taken inside another layer of generic function call.)

  This is necessary to ensure existing `unsafe` code doesn't accidentally transmute hashtables in violation of their coherence expectations, i.e. part of version 1 of "the hashtable problem".

  (Writing it out explicitly like above, without indirection through another generic call, also produces the [Warning: `TypeId` of implementation-aware generic discretised using generic type parameters](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#behaviour-changewarning-typeid-of-implementation-aware-generic-discretised-using-generic-type-parameters), since this is almost never what you actually want to observe when making type-keyed collections.)

- In all other cases, the observed `TypeId` identifies the combination of the type given as type argument and any bounds-relevant captured implementations.

  This makes type-keyed and type-erased collections work as expected by consumer code, without over-distinguishing entries based on scoped implementations that are irrelevant to their function (which is what I call 'version 2 of "the hashtable problem"').

  You can find more on the reasoning here in the section on [Logical consistency of type-erased collections](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#of-type-erased-collections).

Please let me know if this works as clarification.
I should add a cross-reference to the [Logical consistency](https://github.com/Tamschi/rust-rfcs/blob/scoped_impl_trait_for_type/text/3634-scoped-impl-trait-for-type.md#logical-consistency) explanations here in any case, though.

Reply:
Here is my understanding:
- When a concrete type is used as a type argument, the observed type-identity of the corresponding type parameter in the generic item *changes* depending on the where-bounds present on the type parameter.

This seems strange, to say the least. As an example:
```rust
trait TypeContainer {
  type Type;
}

struct Generic<T>(T) /* where T: Marker */;
impl<T> TypeContainer for Generic<T> {
  type Type= T;
}

trait Marker {}

mod one {
  use impl super::Marker for () {}
  pub type Alias = super::Generic<()>;
}

mod two {
  use impl super::Marker for () {}
  pub type Alias = super::Generic<()>;
}

type X = ();
type Y = <one::Alias as TypeContainer>::Type;
type Z = <two::Alias as TypeContainer>::Type;

// or, equivalently
// type Y = <Generic<() as Marker in one> as TypeContainer>::Type;
// type Z = <Generic<() as Marker in two> as TypeContainer>::Type;
```

Is it then the case that X, Y, and Z are the same if there is no `T: Marker` bound, but distinct if there is one (despite all three types being known to be `()`)?

Reply:
> Or is it the case that; no, all three types are known to be identical _in concrete code_, but _generic_ code defining these three types are not only not known to be identical, but are in fact _known to be distinct_ (as witnessed by runtime type comparison).
> 
> […]

This was my intention, but I failed to consider that generic type parameters can flow back into concrete code transparently like this. I'll have to go over this (and the issue) carefully and see how my ideas hold up and where I need to specify more thoroughly or change something.

(_Tentatively_, I think that the implementations can/should 'fall off' in concrete code so that `X`, `Y`, `Z` and `()` are known to be identical and would yield identical `TypeIds`, but that this would not regress the issue because the different concrete `Generic<…>`s are not at all assignable to each other (regardless of whether the bound is present or not).)

I can't look into this properly quite right now unfortunately, but hopefully I'll manage to do it at some point this week.

Reply:
> Tentatively, I think that the implementations can/should 'fall off' in concrete code so that X, Y, Z and () are statically known to be identical and would/could¹ yield identical TypeIds

In effect, implicitly making the type of a generic type parameters a sort of `Wrapper<T>` with partial implementation-awareness.

On a related note, I wonder if you can automate implementation-awareness based on `impl` blocks (e.g. `HashMap` relies on `K: Eq + Hash` in one of its impl blocks, so functions from that impl block carry implementation-awareness, rather than the type itself?

Reply:
> > Tentatively, I think that the implementations can/should 'fall off' in concrete code so that X, Y, Z and () are statically known to be identical and would/could¹ yield identical TypeIds
> 
> In effect, implicitly making the type of a generic type parameters a sort of `Wrapper<T>` with partial implementation-awareness.

In a way. I'll think about whether I explain it that way to some extent, though it's possible it's too not-quite-the-same (for me) to fit it into the text well.

> On a related note, I wonder if you can automate implementation-awareness based on `impl` blocks (e.g. `HashMap` relies on `K: Eq + Hash` in one of its impl blocks, so functions from that impl block carry implementation-awareness, rather than the type itself?

Implementation-awareness must be shared between distinct impl blocks.

For example, the `std`'s `HashMap` has inherent, `Extend`, `From`, `FromIterator`, `Index`, `PartialEq` and `Eq` `impl`s that expect `K: Hash` to be coherent. The category likely also should include external implementations in at least some cases, which for simplicity would mean all cases.

Having that information "flow backwards" from `impl`s to overall `type`s' identities then would introduce a lot of tricky behaviour impacts at a distance, and most likely wouldn't be compatible with early code generation for concrete code.